include/zephyr/sys/util.h

```diff
@@ -129,6 +129,29 @@ extern "C" {
 
 #endif /* __cplusplus */
 
+/**
+ * @brief Declare a flexible array member.
+ *
+ * This macro declares a flexible array member in a struct. The member
+ * is named @p name and has type @p type.
+ *
+ * Since C99, flexible arrays are part of the C standard, but for historical
+ * reasons many places still use an older GNU extension that is declare
+ * zero length arrays.
+ *
+ * Although zero length arrays are flexible arrays, we can't blindly
+ * replace [0] with [] because of some syntax limitations. This macro
+ * workaround these limitations.
+ *
+ * It is specially useful for cases where flexible arrays are
+ * used in unions or are not the last element in the struct.
+ */
+#define FLEXIBLE_ARRAY_DECLARE(type, name) \
+	struct { \
+		struct { } __unused_##name; \
+		type name[]; \
```

Comment:
Suggestion that I missed earlier:
```suggestion
#define FLEXIBLE_ARRAY_DECLARE(_type, _name) \
	struct { \
		struct { } __unused_##name; \
		type name[]; \
```

`type` and `name` are pretty common names, so suggest to add the underscore (commonly used in macro) to avoid any name clashes